nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -0,0 +1,573 @@
+
+  // Using curvature (k = 1 / R) and carrot distance (chord on circle R), project the command
+  const double chord_len = hypot(carrot_pose.pose.position.x, carrot_pose.pose.position.y);
+
+  // Find the number of increments by finding the arc length of the chord on circle
+  const double r = fabs(1.0 / curvature);
+  const double alpha = 2.0 * asin(chord_len / (2 * r)); // central angle
+  const double arc_len = alpha * r;
+  const double delta_dist = costmap_->getResolution();
+  const unsigned int num_pts = static_cast<unsigned int>(ceil(arc_len / delta_dist));
+  const double projection_time = costmap_->getResolution() / linear_vel;
+
+  geometry_msgs::msg::Pose2D curr_pose;
+  curr_pose.x = robot_pose.pose.position.x;
+  curr_pose.y = robot_pose.pose.position.y;
+  curr_pose.theta = tf2::getYaw(robot_pose.pose.orientation);
+
+  for (unsigned int i = 1; i < num_pts; i++) {
+    // only forward simulate within time requested
+    if (i * projection_time > max_allowed_time_to_collision_) {
+      break;
+    }
+
+    // apply velocity at curr_pose over distance delta_dist
+    curr_pose.x += projection_time * (linear_vel * cos(curr_pose.theta));
+    curr_pose.y += projection_time * (linear_vel * sin(curr_pose.theta));
+    curr_pose.theta += projection_time * angular_vel;
+
+    // store it for visualization
+    pose_msg.pose.position.x = curr_pose.x;
+    pose_msg.pose.position.y = curr_pose.y;
+    pose_msg.pose.position.z = 0.01;
+    arc_pts_msg.poses.push_back(pose_msg);
+
+    // check for collision at this point
+    if (inCollision(curr_pose.x, curr_pose.y)) {
+      carrot_arc_pub_->publish(arc_pts_msg);
+      return true;
+    }
+  }
+
+  carrot_arc_pub_->publish(arc_pts_msg);
+
+  return false;
+}
+
+bool RegulatedPurePursuitController::inCollision(const double & x, const double & y)
+{
+  unsigned int mx, my;
+  costmap_->worldToMap(x, y, mx, my);
+
+  unsigned char cost = costmap_->getCost(mx, my);
+
+  if (costmap_ros_->getLayeredCostmap()->isTrackingUnknown()) {
+    return cost >= INSCRIBED_INFLATED_OBSTACLE && cost != NO_INFORMATION;
+  } else {
+    return cost >= INSCRIBED_INFLATED_OBSTACLE;
+  }
+}
+
+double RegulatedPurePursuitController::costAtPose(const double & x, const double & y)
+{
+  unsigned int mx, my;
+  costmap_->worldToMap(x, y, mx, my);
+
+  unsigned char cost = costmap_->getCost(mx, my);
+  return static_cast<double>(cost);
+}
+
+void RegulatedPurePursuitController::applyConstraints(
+  double & linear_vel,
+  const double & dist_error, const double & lookahead_dist,
+  const double & curvature, const geometry_msgs::msg::Twist & curr_speed,
+  const double & pose_cost)
+{
+  // limit the linear velocity by curvature
+  const double radius = fabs(1.0 / curvature);
+  const double & min_rad = regulated_linear_scaling_min_radius_;
+  if (use_regulated_linear_velocity_scaling_ && radius < min_rad) {
+    linear_vel *= 1.0 - (fabs(radius - min_rad) / min_rad);
```

Comment:
That is not the same thing

Reply:
We're calculating an error percentage and inverting it

Reply:
```
1.0 - (fabs(radius - min_radius)/ min_radius)
= 1.0 - ((min_radius - radius)/ min_radius) -> Since radius < min_radius
= (min_radius - (min_radius - radius)) / min_radius
= radius / min_radius
```
Won't this simplification hold?